### PR TITLE
[SPARK-43934][SQL][PYTHON][CONNECT] Add regexp_* functions to Scala and Python

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR want add regexp_* functions to Scala, Python and Connect API.
These functions show below.

- rlike

- regexp

- regexp_count

- regexp_extract_all

- regexp_instr

- regexp_like

- regexp_substr

### Why are the changes needed?
Add regexp_* functions to Scala, Python and Connect API.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New test cases.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @HyukjinKwon @zhengruifeng cc @cloud-fan

File: python/pyspark/sql/connect/functions.py [hunk 1]
```diff
@@ -1988,13 +1988,70 @@ def split(str: "ColumnOrName", pattern: str, limit: int = -1) -> Column:
 split.__doc__ = pysparkfuncs.split.__doc__
 
 
+def rlike(string: "ColumnOrName", pattern: Union[str, Column]) -> Column:
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/functions.py:1991]: for the python side, let's use `ColumnOrName` augments. that is, when `pattern` is a `str`, treat it as a column name

see discussion here https://github.com/apache/spark/pull/41505#discussion_r1222631192

**Comment by zhengruifeng** [comment]
looks pretty good, let's wait for the discussion https://github.com/apache/spark/pull/41516#discussion_r1225030166 , since `regexp_like` and `regexp` are alias to `rlike`

**Comment by beliefer** [comment]
@HyukjinKwon @zhengruifeng The GA failure is unrelated to this PR.

**Review by zhengruifeng — APPROVED** [review]
would you mind rebasing to the lasted master?

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -7596,6 +7721,48 @@ def regexp_extract(str: "ColumnOrName", pattern: str, idx: int) -> Column:
     return _invoke_function("regexp_extract", _to_java_column(str), pattern, idx)
 
 
+@try_remote_functions
+def regexp_extract_all(
+    str: "ColumnOrName", regexp: "ColumnOrName", idx: Optional[Union[int, Column]] = None
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions.py:7726]: does it make sense to make `idx` also support column name?

Comment (beliefer) [reply python/pyspark/sql/functions.py:7726]: I don't quite confirm but according to @HyukjinKwon 's suggestion.

**Comment by beliefer** [comment]
> would you mind rebasing to the lasted master?

Has been rebased. Please wait for GA passed.

**Comment by zhengruifeng** [comment]
merged to master

**Comment by beliefer** [comment]
@zhengruifeng Thank you very much!

**Pull request closed without merging**